File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

/*알고리즘
- 도착 시간을 기준 작업 정렬
- 우선 순위 큐로 작업을 소요 시간 순으로 처리
- 현재 시간 : 도착한 작업은 큐에 넣고, 큐에서 가장 소요 시간이 짧은 작업을 꺼냄
- 대기 시간 : 각 작업 대기 시간 구해서 합평균으로 대기 시간 반환
*/

class Solution {
    public int solution(int[][] jobs) {
        Arrays.sort(jobs, (a, b) -> Integer.compare(a[0], b[0]));

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));

        int time = 0;
        int total = 0;
        int count = 0;
        int index = 0;

        while (count < jobs.length) {
            while (index < jobs.length && jobs[index][0] <= time) {
                pq.offer(jobs[index++]);
            }

            if (pq.isEmpty()) {
                time = jobs[index][0];
            } else {
                int[] current = pq.poll();
                time += current[1];
                total += time - current[0];
                count++;
            }
        }
        int answer = total / jobs.length;
        
        return answer;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
# 알고리즘
# 0. s : 요청 시간, l : 작업 소요 시간, i번 작업
# 1. jobs 리스트 오름차순 정렬하기(요청 시간 기준)
# 2. 힙(우선순위 큐) 사용
# 2-1. (소요시간, 요청 시간)으로 최소힙에 저장 : heapq 이용
# 2-2. 현재 시간(time)을 기준으로 요청 시간 <= 현재 시간인 작업을 힙에 넣기
# 3. 총 걸린 시간을 모두 더해, 작업 개수로 나누기(정수 나누기 '//' 사용)


import heapq

def solution(jobs):
    jobs.sort()
    heap = []
    time, i, total = 0, 0, 0
    count = len(jobs)

    while i < count or heap:
        while i < count and jobs[i][0] <= time:
            heapq.heappush(heap, (jobs[i][1], jobs[i][0]))
            i += 1
            
        if heap:
            dur, req = heapq.heappop(heap)
            time += dur
            total += time - req
        else:
            time = jobs[i][0]
    return total // count